# RNASeq analysis steps
Steps: | Trim | Align | Count | DE | QC | Enrich |
|---|---|---|---|---|---|---|
Tools: | Trimmomatic | STAR | HTSeq | DESeq2 | sklearn & seaborn | clusterprofiler & enrichplot |

# Steps

- Create metadata file
- Create conda environments
- Run 0_reads_to_counts.py
- Run 1_counts_to_de.py
- Run 2_analyze.py

# Create metadata file

## Rename files according to convention

After downloading raw fastq files generated by the sequencing company, the files need to be renamed based on the following convention if required.

**Filename convention: {libname}\_{biorep}\_{techrep}_{readpairnumber}.{fastq|fq}.gz** 

For example, if the library name is *treatment*, biological replicate name is *B1*, technical replicate name is *T1* and the read pair number is *R1*, then the file should be named: *treatment_B1_T1_R1.fastq.gz*

## Create metadata file
Metadata file is a json file containing library specific information along with path to genome fasta file and gtf file. An example metadata file is provided in *example_metadata* folder. Using the same template to create project specific metadata file is recommended.

# Create conda environments

## Conda environment for reads to differential expression
```bash
foo@bar:~$ conda create -n rnaseq -c conda-forge -c anaconda -c bioconda python=3 trimmomatic star htseq bioconductor-deseq2 pandas scikit-learn seaborn -y
```

## Conda environment for enrichment analysis 
```bash
foo@bar$ conda create -n gsea -c bioconda -c conda-forge bioconductor-clusterprofiler bioconductor-biomart bioconductor-org.hs.eg.db bioconductor-enrichplot
```

# Run 0_reads_to_counts.py

# Run 1_counts_to_de.py

# Run 2_analyze.py

**Most tools were chosen following the works of Corchete et. al., Systematic comparison and assessment of RNAâ€‘seq procedures for gene expression quantitative analysis**
